### ISSUE TITLE
problem/bug with cv2.getOptimalNewCameraMatrix

### ISSUE THREAD

**Issue by hilman-dayo** [opencv/opencv#7332]
##### System information (version)

- OpenCV => 3.1
- Operating System / Platform => Mac OS X
##### Detailed description

For OpenCV 3.1 with Python binding, I have done the camera calibration based on this [tutorial](http://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_calib3d/py_calibration/py_calibration.html).
Everything works as expected with the set of pictures that come with OpenCV, but with my own dataset, problem arose with the `cv2.getOptimalNewCameraMatrix`. Basically, it returns a list of ROI of `[0, 0, 0, 0]`. I think maybe it is better if I just give the [link](http://answers.opencv.org/question/102485/problem-with-getoptimalnewcameramatrix/) to the answers.opencv.org where I have asked things regarding this and someone managed to get to the root of the problem. Basically, quoting from the user Tetragramm (the user who has figured out the problem),

> > The why is that the distortion causes the corners of the image to wrap around, making actual negative size ROIs, which get turned into all zeros.

**Comment by kikiorg** [comment]
I think I've figured out what's happening.

**Tl;dr -- make sure the edges are well represented by the checkerboard in your calibration images.**

And so this is not really a fixable bug.

The key is that the coefficients were large numbers, indicating extreme distortion at the edges.  My theory was that without good checkerboard representation at the edges, (given diminishing distortion in the middle) the distortion could solve in widely varying ways, subject to factors like roundoff error and such.

Solving it meant taking calibration images that better represented the edges -- possibly including full frame images.

Longer discussion here: http://answers.opencv.org/question/28438/undistortion-at-far-edges-of-image/?answer=180493#post-id-180493

**Comment by alexsmartens** [comment]
I'm having the same problem. Any ideas?
Thanks!

**Comment by alexsmartens** [comment]
Setting `CALIB_FIX_PRINCIPAL_POINT` flag suggested in #14596 solved the problem for me.

**Comment by rocketsfallonrocketfalls** [comment]
The problem still persists, I've only managed to fix it by looking at @hilman-dayo's thread in https://answers.opencv.org/question/102485/problem-with-getoptimalnewcameramatrix/ and multiplying the focal lengths by 1.01.  

I didn't even know that this was a bug. 5 hours of brainpower has been wasted for nothing. :D
